docs/proposals/autoscaling.md

```diff
@@ -0,0 +1,241 @@
+* auto-scaler - scales replicas up and down by using the `resize` endpoint provided by scalable resources (`ReplicationController`)
+
+
+### Auto-Scaler
+
+The Auto-Scaler is a state reconciler responsible for checking data against configured scaling thresholds 
+and calling the `resize` endpoint to change the number of replicas.  The scaler will 
+use a client/cache implementation to receive watch data from the data aggregators and respond to them by 
+scaling the application.  Auto-scalers are created and defined like other resources via REST endpoints and belong to the 
+namespace just as a `ReplicationController` or `Service`.
+
+There are two options for implementing the auto-scaler:
+
+1.  Annotations on a `ReplicationController`
+    
+    Pros:
+        
+      * uses an existing resource, not another component that must be defined separately
+      * easy to know what the target of the auto-scaler is since the config for the scaler is attached to the target      
+      
+    Cons:
+    
+      * configuration in annotations is marginally more difficult than plain old json  
+      * rather than watching explicitly for new auto-scaler definitions, the auto-scaler controller must watch all 
+      `ReplicationController`s and create auto-scalers when appropriate.  As new auto-scalable resources are defined the 
+      auto-scaler controller must also watch those resources
+      
+1.  As a new resource
+    
+    Pros:
+        
+      * auto-scalers are managed by the user independent of the `ReplicationController`
+      * flexible by using a selector to the scalable resource (that implements the `resize` verb), future implementations 
+      *may* require no extra work on the auto-scaler side
+      
+    Cons:
+    
+      * one more resource to store, manage, and monitor
+  
+For this proposal, the auto-scaler is a resource:
+
+    //The auto scaler interface
+    type AutoScalerInterface interface {        
+        //Adjust a resource's replica count.  Calls resize endpoint.  Args to this are based on what the endpoint
+        //can support.  See https://github.com/GoogleCloudPlatform/kubernetes/issues/1629
+        ScaleApplication(num int) error
+    }
+
+    type AutoScaler struct {
+        //common construct
+        TypeMeta
+        //common construct
+        ObjectMeta
+        
+        //Spec defines the configuration options that drive the behavior for this auto-scaler 
+        Spec    AutoScalerSpec  
+        
+        //Status defines the current status of this auto-scaler.
+        Status  AutoScalerStatus               
+     }
+     
+    type AutoScalerSpec struct {
+        //Thresholds
+        AutoScaleThresholds []AutoScaleThreshold
+        
+        //turn auto scaling on or off
+        Enabled boolean 
+        
+        //max replicas that the auto scaler can use, empty is unlimited
```

Comment:
Maxint? :-)  0 is a valid number of replicas, but could maybe be banned in MaxAutoScaleCount, since there is a separate Enabled field. Or, the field could just be made required and the user could specify a fairly large number.